Sprint 16

v-alekseev opened the pull request.

v-alekseev committed:
Реализован онбординг и работа с категориями.  SelectGroupViewController переведен на MVVM архитектуру

v-alekseev committed:
Случайно не закоммитился файл

v-alekseev committed:
Мелкие правки, удаление отладочной печати

LizaIlinykh commented on Tracker/CoreData/TrackerCategoryStore.swift:
```diff
@@ -9,12 +9,18 @@ import Foundation
 final class TrackerCategoryStore: NSObject {
     
+    
+    var delegate: TrackerCategoryStoreDelegateProtocol?
```
слабая ссылка на делегат нужна

LizaIlinykh commented on Tracker/Models/TrackersModels.swift:
```diff
@@ -10,20 +10,20 @@ import UIKit
+//    static func trackerIDsFromString(udids: String) -> [UUID] {
+//        let udidStingArray = udids.split(separator: ",")
+//        return udidStingArray.compactMap{ UUID(uuidString: String($0))}
+//    }
```
если закомментированный код не нужен, его лучше удалить

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+}
+
+
+class OnboardingViewController: UIPageViewController, UIPageViewControllerDataSource, UIPageViewControllerDelegate {
```
final

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+
+class OnboardingViewController: UIPageViewController, UIPageViewControllerDataSource, UIPageViewControllerDelegate {
+
+    lazy var pages: [UIViewController] = {
```
private

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+extension UIPageControl {
+    
+    override open func sizeThatFits(_ size: CGSize) -> CGSize {
+        print("sizeThatFits")
```
можно удалить принт

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        return [first, second]
+    }()
+    
+    func setUPPage(imageName: String, text: String) -> UIViewController {
```
private func setupPage

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        labelText.centerXAnchor.constraint(equalTo: viewController.view.centerXAnchor).isActive  = true
+        labelText.topAnchor.constraint(equalTo: viewController.view.topAnchor, constant: 432).isActive = true
+        labelText.leadingAnchor.constraint(equalTo: viewController.view.leadingAnchor, constant: 16).isActive = true
+        labelText.trailingAnchor.constraint(equalTo: viewController.view.trailingAnchor, constant: -16).isActive = true
```
активировать констрейнты можно с помощью функции NSLayoutConstraint.activate([консрейнты которые нужно активировать])

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        return viewController
+    }
+
+    lazy var pageControl: UIPageControl = {
```
private

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+        return pageControl
+    }()
+    
+    lazy var buttonOnboardinComplete: UIButton = {
```
private

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+    }()
+
+
+    @objc func buttonOnboardinCompleteTapped() {
```
private

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+    // MARK: - UIPageViewControllerDataSource
+
+    // предыдущий экран
+    func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
```
советую выносить методы с реализацией протоколов в отдельные расширения

v-alekseev replied:
В данном случае лучше в одном файла. Крайности всегда плохо.

LizaIlinykh replied:
я не говорила про файлы, можно вынести в расширение. обычно такой формат используют для описания реализации протокола

v-alekseev replied:
Это да, поправил.  Но изначально  копировал файл из лекции.  протоколы через запятую  определении класса

LizaIlinykh commented on Tracker/Onboarding/OnboardingViewController.swift:
```diff
@@ -0,0 +1,179 @@
+            pageControl.currentPage = currentIndex
+        }
+    }
+}
```
в это классе нарушен порядок элементов
вот ссылка для ознакомления, соблюдение общепринятого формата сделает твой код более удобным для чтения и в нем будет легче ориентироваться
https://habr.com/ru/articles/744470/

LizaIlinykh commented on Tracker/ViewControllers/Category/CreateCategoryViewController.swift:
```diff
@@ -0,0 +1,140 @@
+    // MARK: public properties
+    var selectGroupViewModel: SelectGroupViewModel?
+    
+    let textBackgroundView: UIView = {
```
все вью нужно делать приватными, если их свойства хочется менять снаружи, стоит это делать с помощью методов

LizaIlinykh commented on Tracker/ViewControllers/Category/CreateCategoryViewController.swift:
```diff
@@ -0,0 +1,140 @@
+    }()
+    
+    // MARK: - UI elemants
+    lazy private var categoryNameTextView: UITextField = {
```
сверху тоже юай элемент

LizaIlinykh commented on Tracker/ViewControllers/Category/CreateCategoryViewController.swift:
```diff
@@ -0,0 +1,140 @@
+    }()
+    
+    
+    @objc func buttonCreateCategoryTapped() {
```
private

LizaIlinykh commented on Tracker/ViewControllers/Category/CreateCategoryViewController.swift:
```diff
@@ -0,0 +1,140 @@
+            if selectGroupViewModel.addCategory(name:  text) == false {
+                Alert.alertInformation(viewController: self, text: "Ошибка создания категории") }
+        }
+        
```
немного поправила кодстайл
```
    @objc func buttonCreateCategoryTapped() {
        guard let selectGroupViewModel = selectGroupViewModel,
              let text = categoryNameTextView.text else { return }
        
        if isUpdateAction {
            if !selectGroupViewModel.renameCategory(name:  currentCategoryName ?? "",
                                                    newCategoryName: text) {
                Alert.alertInformation(viewController: self, text: "Ошибка редактирования категории")
            }
        } else {
            if !selectGroupViewModel.addCategory(name:  text) {
                Alert.alertInformation(viewController: self, text: "Ошибка создания категории")
            }
        }
        
        dismiss(animated: true)
    }
```

а еще лучше выкидывать ошибки вместо получения true / false
потому что иногда нам нужно знать, что за ошибка произошла и обработать ее особенным образом

v-alekseev replied:
Да,  конечно лучше детализировать причины ошибки. Через throw или еще как, не так важно. Но времени на этот рефакторинг нет совсем(((

LizaIlinykh commented on Tracker/ViewControllers/Category/CreateCategoryViewController.swift:
```diff
@@ -0,0 +1,140 @@
+        dismiss(animated: true)
+    }
+    
+    @objc func labelTextChanged() {
```
private

LizaIlinykh commented on Tracker/ViewControllers/Category/CreateCategoryViewController.swift:
```diff
@@ -0,0 +1,140 @@
+        ])
+    }
+    
+}
```
в это классе также нарушен порядок элементов

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    // MARK: Constans
+    private let cellID = "cell"
+    // MARK: public properties
+    var createTrackerViewController: CreateTrackerViewController?
```
это тоже делегат, должен быть слабым

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    // MARK: public properties
+    var createTrackerViewController: CreateTrackerViewController?
+    
+    var selectGroupViewModel = SelectGroupViewModel()
```
лучше не создавать зависимости внутри классов, а внедрять их извне. 
вот ссылка на статью о внедрении зависимостей
https://swiftblog.org/vnedrenie-zavisimostej/

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+
+
+final class SelectGroupViewController: UIViewController {
+    
```
одной пустой строки для отступов достаточно

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    var selectGroupViewModel = SelectGroupViewModel()
+    
+    // MARK: - UI elemants
+    //private var createNewGroupButton = UIButton()
```
не нужный код стоит удалить

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+        return noCategoryLabel
+    }()
+    
+    lazy var createGroupButton: UIButton = {
```
private

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    }()
+    
+    
+    @objc func buttonCreateNewCategoryTapped() {
```
private

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    @objc func buttonCreateNewCategoryTapped() {
+        let createGroupViewController = CreateGroupViewController(isUpdateAction: false)
+        
+        createGroupViewController.selectGroupViewModel = self.selectGroupViewModel
```
не нужно использовать для двух экранов одну вью модель
для каждого экрана нужно создать своб собственную вью модель, для соблюдения single responsibility

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+        
+        let navigationController = UINavigationController(rootViewController: createGroupViewController)
+        navigationController.modalPresentationStyle = .pageSheet
+        self.present(navigationController, animated: true)
```
также для четкого распределния обязанностей рекомендую создать фабрику , чтобы она занималась сборкой модулей
также можно создать роутер , чтоб он занимался навигацией

это все не обязанности вью

v-alekseev replied:
1. роутер - это часть архитектуры VIPER?
2.  "Рекомендую вынести сборку модулей в фабрику" - а что подразумевается под сборкой модулей? можно пример или раскрыть фразу.

LizaIlinykh replied:
роутер - это сущность ответственная за навигацию. в архитектуре вайпер действительно есть роутер, но роутер может использоваться где угодно. Выглядит примерно так:
```

final class Router: RouterProtocol {

    private weak var viewController: ViewController?
    
    init(viewController: ViewController) {
        self.viewController = viewController
    }
    
        func goToSomeScreen() {
                guard let navigationController = viewController?.navigationController else { return }
                navigationController.popViewController(animated: true)
        }
}

```

LizaIlinykh replied:
один модуль  в архитектуре mvvm - это вью + вью модель, если будет еще роутер, то он тоже будет частью модуля. сборка этого модуля тоже отдельная отвественность которую можно вынести в отдельную сущность, есть ассемблеры например. но я предлагаю сделать фабрику

```
struct ModuleFactory {
     static func createSelectGroupModule() -> ViewController {
           let vc = ViewController()
           let viewModel = ViewModel()
           vc.viewModel = viewModel
           /// если сделаешь роутер 
          let router = Router()
          router.viewController = vc
         return vc
     }
}

```

это сейчас кажется лишним выносить это в отдельные сущности, но потом появятся разные сервисы и менеджеры которые надо будет прописывать в каждой части программы которая создает этот экран. а если всего одна функция этот экран собирает - нужно будет поправить сборку только в одном месте

v-alekseev replied:
Спасибо большое!

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+        return menu
+    }
+    
+}
```
в этом классе тоже нарушен порядок элементов, стоит поправить

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+            self?.groupsTable.reloadData()
+        }
+        
+        selectGroupViewModel.selectGroupViewController = self
```
советую вынести все эти настройки в отдельные методы, чтобы во viewdidload не разрасталась логика

v-alekseev replied:
Ту всего несколько строк. Плодить функции тоже  плохо.

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+    }
+    
+    // MARK: Private functions
+    func alert(text: String) {
```
private

v-alekseev replied:
alet вызывается из extension. она должна быть публичная

LizaIlinykh replied:
тогда стоит убрать пометку Private functions

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewController.swift:
```diff
@@ -0,0 +1,233 @@
+        Alert.alertInformation(viewController: self, text: text)
+    }
+    
+    func showLogo(_ uiShow: Bool) {
```
private

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+
+extension SelectGroupViewModel: TrackerCategoryStoreDelegateProtocol {
+    func didUpdate() {
+        guard let a = trackerCategoryStore.getCategories() else { return }
```
a - плохое имя для константы

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+    }
+    
+    func renameCategory(name: String, newCategoryName: String)  -> Bool {
+        if(name == "") {
```
guard !name.isEmpty else { return false }

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+    
+    //TODO: did set setCategoryName
+    
+    private var trackerCategoryStore = TrackerCategoryStore()
```
лучше заносить через инициализатор
выше кидала ссылку по внедрению зависимостей

v-alekseev replied:
В лекцих во всех примерах было так.
Что бы занести через внедрения зависимостей, нужно обернуть это в протокол.  Может попозже сделаю, пока не до этого.

LizaIlinykh commented on Tracker/ViewControllers/Category/SelectGroupViewModel.swift:
```diff
@@ -0,0 +1,58 @@
+    init() {
+        trackerCategoryStore.delegate = self
+        
+        guard let a = trackerCategoryStore.getCategories(),
```
a - плохое имя для константы

LizaIlinykh commented on Tracker/ViewControllers/CreateTrackerView/CollectionsDelegate/ColorViewControllerDelegate.swift:
```diff
@@ -65,7 +65,7 @@ extension ColorViewControllerDelegate: UICollectionViewDelegate {
     // размер ячейки
     func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
-        var cellWight = (collectionView.bounds.width-18-19-(5*5))/6
+        let cellWight = (collectionView.bounds.width-18-19-(5*5))/6
```
не стоит использовать магические числа
https://ru.hexlet.io/blog/posts/magic-numbers

LizaIlinykh commented on Tracker/ViewControllers/CreateTrackerView/CreateTrackerViewController.swift:
```diff
@@ -63,7 +62,18 @@ final class CreateTrackerViewController: UIViewController {
     
     private var scrollView = UIScrollView()
```
let

LizaIlinykh commented on Tracker/ViewControllers/TrackersView/TrackersViewController.swift:
```diff
@@ -128,26 +136,25 @@ class TrackersViewController: UIViewController {
+                
             } else {
                 let a = tracker.trackerActiveDays.filter { DaysOfWeek(rawValue: $0 ) == currentDayOfWeek }
                 if a.count > 0 { // один из дней совпас с сегодняшним днкм недели
```
имя a  - плохое имя для константы

LizaIlinykh commented on Tracker/ViewControllers/TrackersView/extensions/TrackersViewController+UICollectionViewDataSource.swift:
```diff
@@ -12,13 +12,16 @@ import UIKit
-        let tracker =  visibleTrackers[indexPath.row] //   trackers[indexPath.row]
+        
+        let categoryName = visibleTrackers.getKeyByIndex(index: indexPath.section)
+        let tracker = visibleTrackers[categoryName]![indexPath.row]
```
давай избавимся от принудительного извлечения

v-alekseev committed:
- Вместо стандартной ячейки UITableViewCell используется кастомная SelectGroupTableViewCell.
- В последней ячейки таблицы добавлено закругление.
- Сепаратор в последней ячейке теперь не отображается.

v-alekseev committed:
На экране создания трекера, когда категория не выбрана лейбл Категория теперь по середине

v-alekseev committed:
- Для создания и для редактирования категории теперь используются отдельные контроллеры EditCategoryViewController и CreateCategoryViewController
- Убраны "магические" цифры

v-alekseev committed:
- Вью модель теперь обрабатывать любое действие пользователя, в том числе нажатие на ячейку
- в класс ячейки перенес функции из VC  по настройки ее отображения
- выделил делегаты в отдельные файлы

v-alekseev committed:
extension вынесены в отдельные файлы

v-alekseev committed:
Исправлены замечания из github

v-alekseev committed:
- Созданы отдельные  ViewModel для экранов создания и редактирования категорий
- Обработка нажатия на таблицу перенесена в SelectGroupViewModel

v-alekseev committed:
Переделана структура файлов и EditCategory и CreateCategory доведены до MVVM архитектуры

v-alekseev commented:
Я в первый раз перепутал  что надо разделить. Я разделил ViewController  экрана редактирование/создание на 2 контроллера (редактирование и создание) (((. А надо было ViewModel разделять(
Сейчас разделил SelectGroupViewModel на SelectGroupViewModel, CreateGroupViewModel и EditCategoryViewModel.
А еще пришлось добавить NSFetchedResultsController что бы ловить обновление модели из  CreateGroupViewModel и EditCategoryViewModel.

v-alekseev commented:
Экраны создания и редактирования обратно в один экран не стал сливать уже.

v-alekseev commented:
Нижний сепаратор в другой таблице уберу позжу

v-alekseev commented:
Router тоже надо сделать, но после сдачи спринта. Иначе все развалю в внесу кучу багов.

LizaIlinykh commented on Tracker/ViewControllers/Category/CreateCategoryScreen/CreateCategoryViewController.swift:
```diff
@@ -10,20 +10,9 @@ import UIKit
 
     // MARK: public properties
-    private var selectGroupViewModel: SelectGroupViewModel
+    private var createGroupViewModel = CreateGroupViewModel()
```
let

LizaIlinykh commented on Tracker/ViewControllers/Category/CreateCategoryScreen/CreateCategoryViewController.swift:
```diff
@@ -86,7 +75,7 @@ final class CreateGroupViewController: UIViewController {
         guard let text = categoryNameTextView.text else { return }
         
-        if selectGroupViewModel.addCategory(name:  text) == false {
+        if createGroupViewModel.addCategory(name:  text) == false {
```
if !createGroupViewModel.addCategory(name:  text) {

LizaIlinykh commented on Tracker/ViewControllers/Category/EditCategoryScreen/EditCategoryViewModel.swift:
```diff
@@ -0,0 +1,31 @@
+    func renameCategory() {
+        guard !newCategoryName.isEmpty else { return }
+        
+        if trackerCategoryStore.updateCategory(category: TrackerCategory(categoryName: currentCategoryName), newCategory: TrackerCategory(categoryName: newCategoryName)) == false {
```
!trackerCategoryStore.updateCategory(category: TrackerCategory(categoryName: currentCategoryName), newCategory: TrackerCategory(categoryName: newCategoryName))
